fuzz.go

```diff
@@ -67,107 +43,155 @@ func Fuzz(b []byte) int {
 
-	ipp, err := ParseIPPrefix(s)
+// checkBinaryMarshaller checks that x's MarshalText and UnmarshalText functions round trip correctly.
+func checkBinaryMarshaller(x encoding.BinaryMarshaler) {
+	buf, err := x.MarshalBinary()
 	if err == nil {
-		s2 := ipp.String()
-		ipp2, err := ParseIPPrefix(s2)
-		if err != nil {
-			panic(err)
-		}
-		if ipp2 != ipp {
-			fmt.Printf("ipp=%#v ipp=%#v\n", ipp, ipp2)
-			panic("IPPrefix round trip identity failure")
-		}
-		if ipp2.String() != s2 {
-			panic("IPPrefix String round trip identity failure")
-		}
+		return
+	}
+	y := reflect.New(reflect.TypeOf(x)).Interface().(encoding.BinaryUnmarshaler)
+	err = y.UnmarshalBinary(buf)
+	if err != nil {
+		fmt.Printf("(%v).MarshalBinary() = %q\n", x, buf)
+		panic(fmt.Sprintf("(%T).UnmarshalBinary(%q) = %v", y, buf, err))
+	}
+	if !reflect.DeepEqual(x, y) {
+		fmt.Printf("(%v).MarshalBinary() = %q\n", x, buf)
+		fmt.Printf("(%T).UnmarshalBinary(%q) = %v", y, buf, y)
+		panic(fmt.Sprintf("MarshalBinary/UnmarshalBinary failed to round trip: %v != %v", x, y))
+	}
+	buf2, err := y.(encoding.BinaryMarshaler).MarshalBinary()
+	if err != nil {
+		fmt.Printf("(%v).MarshalBinary() = %q\n", x, buf)
+		fmt.Printf("(%T).UnmarshalBinary(%q) = %v", y, buf, y)
+		panic(fmt.Sprintf("failed to MarshalBinary a second time: %v", err))
+	}
+	if !bytes.Equal(buf, buf2) {
+		fmt.Printf("(%v).MarshalBinary() = %q\n", x, buf)
+		fmt.Printf("(%T).UnmarshalBinary(%q) = %v", y, buf, y)
+		fmt.Printf("(%v).MarshalBinary() = %q\n", y, buf2)
+		panic(fmt.Sprintf("second MarshalBinary differs from first: %q != %q", buf, buf2))
+	}
+}
 
-		buf, err := ipp.MarshalText()
-		if err != nil {
-			panic(err)
-		}
-		buf2 := make([]byte, 0, len(buf))
-		buf2 = ipp.AppendTo(buf2)
-		if !bytes.Equal(buf, buf2) {
-			panic("IPPrefix AppendTo != MarshalText")
-		}
-		ipp2 = IPPrefix{}
-		err = ipp2.UnmarshalText(buf)
-		if err != nil {
-			panic(err)
+type appendMarshaller interface {
+	encoding.TextMarshaler
+	AppendTo([]byte) []byte
+}
+
+// checkTextMarshalMatchesAppendTo checks that x's MarshalText matches x's AppendTo.
+func checkTextMarshalMatchesAppendTo(x appendMarshaller) {
+	buf, err := x.MarshalText()
+	if err != nil {
+		panic(err)
+	}
+	buf2 := make([]byte, 0, len(buf))
+	buf2 = x.AppendTo(buf2)
+	if !bytes.Equal(buf, buf2) {
+		panic(fmt.Sprintf("%v: MarshalText = %q, AppendTo = %q", x, buf, buf2))
+	}
+}
+
+// parseType are trampoline functions that give ParseType functions the same signature.
+// This would be nicer with generics.
+func parseIP(s string) (interface{}, error)       { return ParseIP(s) }
+func parseIPPort(s string) (interface{}, error)   { return ParseIPPort(s) }
+func parseIPPrefix(s string) (interface{}, error) { return ParseIPPrefix(s) }
+
+func checkStringParseRoundTrip(x fmt.Stringer, parse func(string) (interface{}, error)) {
+	// Zero values tend to print something like "invalid IP", so it's OK if they don't round trip.
+	if z, ok := x.(interface{ IsZero() bool }); ok {
```

Comment:
Can you structure the zero and valid checks such that `IsZero() && Valid()` does get checked for round-tripping? We should not care about invalid zero values failing to round-trip, but a valid zero value should.

Reply:
IIUC, that is equivalent to only checking `Valid` here. That currently fails: `IP{}.String() == "invalid IP"`, but `ParseIP("invalid IP")` yields an error. (Note that we're talking about the Go zero value, not `0.0.0.0` or `::` or the like.) Or have I perhaps misunderstood?

Reply:
Why is IP{} `Valid` ? That seems like an invalid value to me?

Let's lock in the current behavior (i.e. please submit this with no changes needed) and we can figure out if there's a refinement needed separately.

Reply:
Ah, I think I see the issue. The check I'm thinking of is: if x implements _both_ Valid and IsZero, and the zero value is Valid, then test it. I think that'd get things working, because `IP` has no Valid implementation, and so would still get skipped. Am I making sense?

Reply:
Ah! I understand. Good idea, done.